src/hierarchies/mfa_hierarchy.cc

```diff
@@ -0,0 +1,288 @@
+
+  else {
+    throw std::invalid_argument("Unrecognized hierarchy prior");
+  }
+}
+
+void MFAHierarchy::update_summary_statistics(const Eigen::RowVectorXd& datum,
+                                             bool add) {
+  if (add) {
+    data_sum += datum;
+    data.push_back(datum);
+  } else {
+    data_sum -= datum;
+    auto pos = std::find(data.begin(), data.end(), datum);
+    data.erase(pos);
+  }
+}
+
+void MFAHierarchy::clear_summary_statistics() {
+  data_sum = Eigen::VectorXd::Zero(p);
+  data.clear();
+}
+
+void MFAHierarchy::set_state_from_proto(
+    const google::protobuf::Message& state_) {
+  auto& statecast = downcast_state(state_);
+  state.mu = bayesmix::to_eigen(statecast.mfa_state().mu());
+  state.psi = bayesmix::to_eigen(statecast.mfa_state().psi());
+  state.Eta = bayesmix::to_eigen(statecast.mfa_state().eta());
+  state.Lambda = bayesmix::to_eigen(statecast.mfa_state().lambda());
+  set_card(statecast.cardinality());
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::ClusterState>
+MFAHierarchy::get_state_proto() const {
+  bayesmix::MFAState state_;
+  bayesmix::to_proto(state.mu, state_.mutable_mu());
+  bayesmix::to_proto(state.psi, state_.mutable_psi());
+  bayesmix::to_proto(state.Eta, state_.mutable_eta());
+  bayesmix::to_proto(state.Lambda, state_.mutable_lambda());
+
+  auto out = std::make_shared<bayesmix::AlgorithmState::ClusterState>();
+  out->mutable_mfa_state()->CopyFrom(state_);
+  return out;
+}
+
+void MFAHierarchy::set_hypers_from_proto(
+    const google::protobuf::Message& hypers_) {
+  auto& hyperscast = downcast_hypers(hypers_).mfa_state();
+  hypers->mutilde = bayesmix::to_eigen(hyperscast.mutilde());
+  hypers->alpha0 = hyperscast.alpha0();
+  hypers->beta = bayesmix::to_eigen(hyperscast.beta());
+  hypers->phi = hyperscast.phi();
+  hypers->q = hyperscast.q();
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::HierarchyHypers>
+MFAHierarchy::get_hypers_proto() const {
+  bayesmix::MFAPriorDistribution hypers_;
+  bayesmix::to_proto(hypers->mutilde, hypers_.mutable_mutilde());
+  bayesmix::to_proto(hypers->beta, hypers_.mutable_beta());
+  hypers_.set_alpha0(hypers->alpha0);
+  hypers_.set_phi(hypers->phi);
+  hypers_.set_q(hypers->q);
+
+  auto out = std::make_shared<bayesmix::AlgorithmState::HierarchyHypers>();
+  out->mutable_mfa_state()->CopyFrom(hypers_);
+  return out;
+}
+
+void MFAHierarchy::sample_full_cond(bool update_params) {
+  assert(update_params != true);  // should never be true
+  if (this->card == 0) {
+    // No posterior update possible
+    sample_prior();
+  } else {
+    sample_Eta();
+    sample_mu();
+    sample_psi();
+    sample_Lambda();
+  }
+}
+
+void MFAHierarchy::sample_Eta() {
+  auto& rng = bayesmix::Rng::Instance().get();
+
+  Eigen::MatrixXd Sigmaeta(
+      (Eigen::MatrixXd::Identity(hypers->q, hypers->q) +
+       state.Lambda.transpose() *
+           Eigen::MatrixXd(state.psi.cwiseInverse().asDiagonal()) *
+           state.Lambda)
+          .inverse());
```

Comment:
Just a reminder that we'll need to come back to this, since we want to avoid inversions as much as possible. Same for other inversions in this class

Reply:
All the inverses are just temporary, we might not even need to invert them, and in case we needed to we already noticed how you do it using Eigen's methods to solve linear systems.